### PR TITLE
[SandboxVec] Boilerplate

### PR DESCRIPTION
**Pull request by vporpo**
This patch implements the new pass and registers it with the pass manager. For context, this is a vectorizer that operates on Sandbox IR, which is a transactional IR on top of LLVM IR.

### CODE REVIEW

**Review by aeubanks — COMMENTED** [review]
I wouldn't shorten "Vec" in the directory name

File: llvm/lib/Transforms/Vectorize/SandboxVec/SandboxVectorizer.cpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+#define DEBUG_TYPE "SBVec"
+
+cl::opt<bool>
+    SBVecDisable("sbvec-disable", cl::init(false), cl::Hidden,
+                 cl::desc("Disable the Sandbox Vectorization passes"));
+
+PreservedAnalyses SandboxVectorizerPass::run(Function &F,
+                                             FunctionAnalysisManager &AM) {
+  TTI = &AM.getResult<TargetIRAnalysis>(F);
+  SE = &AM.getResult<ScalarEvolutionAnalysis>(F);
```
Comment (tschuett) [inline llvm/lib/Transforms/Vectorize/SandboxVec/SandboxVectorizer.cpp:24]: Most of this is dead code.

File: llvm/lib/Transforms/Vectorize/SandboxVec/SandboxVectorizer.cpp [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+#include "llvm/Transforms/Vectorize/SandboxVec/SandboxVectorizer.h"
+#include "llvm/Support/CommandLine.h"
+
+using namespace llvm;
+
+#define SV_NAME "sandbox-vectorizer"
+#define DEBUG_TYPE "SBVec"
+
+cl::opt<bool>
+    SBVecDisable("sbvec-disable", cl::init(false), cl::Hidden,
```
[the remark this answers is not in the document]
Comment (tschuett) [reply llvm/lib/Transforms/Vectorize/SandboxVec/SandboxVectorizer.cpp:18]: As long as you are not using the pass in production, opt should suffice for use it or not.

File: llvm/lib/Transforms/Vectorize/SandboxVec/SandboxVectorizer.cpp (hunk 1 above)
Comment (vporpo) [reply llvm/lib/Transforms/Vectorize/SandboxVec/SandboxVectorizer.cpp:24]: Yeah I guess I can drop these for now and add them when needed.

**Comment by tschuett** [comment]
The vectorizer is going to be spread of several files and some files will be added to the generic sandbox directory?
E.g., the pass manager could be used for other optimizers on top of sandbox.

**Comment by vporpo** [comment]
> The vectorizer is going to be spread of several files and some files will be added to the generic sandbox directory?

Not sure what you mean by `generic sandbox directory`.

There will be a number of SandboxVectorizer-specific files living in `Vectorize/SandboxVectorizer`. This will include several components of the vectorizer, and a list of vectorization passes which will be called by the sandbox-vectorizer speicific pass manager.

**Comment by vporpo** [comment]
> Could the pass manager be used for other optimizations on top of SandboxIR or is it tied to the vectorizer?

The pass manager will be launched by the vectorizer pass, so in that sense it is tied to the vectorizer. But I see your point. Ideally there should be a generic pass manager for Sandbox IR passes. I think we could accommodate for that if Sandbox IR ended up being used by other transformations too. In that case we could move the Pass/PassManager files to a generic directory.

**Comment by tschuett** [comment]
My dream would be a loop optimizer on top of SandboxIR. It would also have passes and a pass manager.

**Comment by tschuett** [comment]
Neither worry nor hurry.

**Comment by vporpo** [comment]
> My dream would be a loop optimizer on top of SandboxIR. It would also have passes and a pass manager.

Yeah, loop transformations would be a great fit. I think the pass-manager infrastructure is generic enough that we could reuse most of it.

**Comment by vporpo** [comment]
I accidentally force-pushed, but I think I addressed all comments.

**Comment by vporpo** [comment]
Ah I still need to rename the test directory from `SandboxVec` to `SandboxVectorizer`

File: llvm/lib/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  return PA;
+}
+
+bool SandboxVectorizerPass::runImpl(Function &F) {
+  if (SBVecDisable)
+    return false;
+
+  // If the target claims to have no vector registers don't attempt
+  // vectorization.
+  if (!TTI->getNumberOfRegisters(TTI->getRegisterClassForType(true))) {
```
Comment (aeubanks) [inline llvm/lib/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.cpp:40]: I'd add this (and below) logic alongside tests later

File: llvm/lib/Transforms/Vectorize/SandboxVec/SandboxVectorizer.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (aeubanks) [reply llvm/lib/Transforms/Vectorize/SandboxVec/SandboxVectorizer.cpp]: it's not being added to any pipelines, so this option isn't useful right now. and once it is added to the pipelines, it'll be controlled by `PipelineTuningOptions::SLPVectorization` anyway

**Comment by vporpo** [comment]
> it's not being added to any pipelines, so this option isn't useful right now. and once it is added to the pipelines, it'll be controlled by PipelineTuningOptions::SLPVectorization anyway

Done.

File: llvm/lib/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.cpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.h"
+#include "llvm/Support/CommandLine.h"
+
+using namespace llvm;
+
+#define SV_NAME "sandbox-vectorizer"
+#define DEBUG_TYPE "SBVec"
```
Comment (alinas) [inline llvm/lib/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.cpp:15]: `SBVectorizer`?

**Review by alinas — APPROVED** [review]

File: llvm/lib/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.cpp (hunk 4 above)
Comment (vporpo) [reply llvm/lib/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.cpp:15]: We could change it to `#define DEBUG_TYPE SV_NAME` like it's done in LoopVectorize.cpp. Wdyt?

Comment (alinas) [reply llvm/lib/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.cpp:15]: SGTM.

File: llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.h [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+//===----------------------------------------------------------------------===//
+#ifndef LLVM_TRANSFORMS_VECTORIZE_SANDBOXVECTORIZER_SANDBOXVECTORIZER_H
+#define LLVM_TRANSFORMS_VECTORIZE_SANDBOXVECTORIZER_SANDBOXVECTORIZER_H
+
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/ScalarEvolution.h"
+#include "llvm/Analysis/TargetTransformInfo.h"
+#include "llvm/IR/DataLayout.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/SandboxIR/SandboxIR.h"
```
Comment (nikic) [inline llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.h:16]: Most of these includes are either (currently) unused or should be forward declarations.

Comment (vporpo) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.h:16]: Yeah I removed the objects themselves and forgot to remove the header files. Let me remove them.

Comment (vporpo) [reply llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/SandboxVectorizer.h:16]: Removed them here: 7ea9f0d85fc3dc80b45e6ba7087c41c6f2481f07

**Comment by mikaelholmen** [comment]
Hi,
I've no idea what to be expected but I noticed that
```
opt -passes=sandbox-vectorizer bbi-98907.ll -o /dev/null
```
crashes with
```
opt: ../lib/SandboxIR/SandboxIR.cpp:2542: llvm::sandboxir::Value *llvm::sandboxir::Context::registerValue(std::unique_ptr<Value> &&): Assertion `Pair.second && "Already exists!"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: build-all/bin/opt -passes=sandbox-vectorizer bbi-98907.ll -o /dev/null
1.	Running pass "function(sandbox-vectorizer)" on module "bbi-98907.ll"
2.	Running pass "sandbox-vectorizer" on function "get"
 #0 0x000055ee9d6fb447 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (build-all/bin/opt+0x419b447)
 #1 0x000055ee9d6f8f2e llvm::sys::RunSignalHandlers() (build-all/bin/opt+0x4198f2e)
 #2 0x000055ee9d6fbc6f SignalHandler(int) Signals.cpp:0:0
 #3 0x00007f128e35fcf0 __restore_rt (/lib64/libpthread.so.0+0x12cf0)
 #4 0x00007f128bf18acf raise (/lib64/libc.so.6+0x4eacf)
 #5 0x00007f128beebea5 abort (/lib64/libc.so.6+0x21ea5)
 #6 0x00007f128beebd79 _nl_load_domain.cold.0 (/lib64/libc.so.6+0x21d79)
 #7 0x00007f128bf11426 (/lib64/libc.so.6+0x47426)
 #8 0x000055eea0de61e7 llvm::sandboxir::Context::registerValue(std::unique_ptr<llvm::sandboxir::Value, std::default_delete<llvm::sandboxir::Value>>&&) (build-all/bin/opt+0x78861e7)
 #9 0x000055eea0de7971 llvm::sandboxir::Context::createBasicBlock(llvm::BasicBlock*) (build-all/bin/opt+0x7887971)
#10 0x000055eea0de8861 llvm::sandboxir::Context::createFunction(llvm::Function*) (build-all/bin/opt+0x7888861)
#11 0x000055ee9f394dab llvm::SandboxVectorizerPass::runImpl(llvm::Function&) (build-all/bin/opt+0x5e34dab)
#12 0x000055ee9f3949fc llvm::SandboxVectorizerPass::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x5e349fc)
#13 0x000055ee9f249f4d llvm::detail::PassModel<llvm::Function, llvm::SandboxVectorizerPass, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilder.cpp:0:0
#14 0x000055ee9d904d6a llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) (build-all/bin/opt+0x43a4d6a)
#15 0x000055ee9ea8a2dd llvm::detail::PassModel<llvm::Function, llvm::PassManager<llvm::Function, llvm::AnalysisManager<llvm::Function>>, llvm::AnalysisManager<llvm::Function>>::run(llvm::Function&, llvm::AnalysisManager<llvm::Function>&) PassBuilderPipelines.cpp:0:0
#16 0x000055ee9d909901 llvm::ModuleToFunctionPassAdaptor::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x43a9901)
#17 0x000055ee9ea837fd llvm::detail::PassModel<llvm::Module, llvm::ModuleToFunctionPassAdaptor, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) PassBuilderPipelines.cpp:0:0
#18 0x000055ee9d9039fa llvm::PassManager<llvm::Module, llvm::AnalysisManager<llvm::Module>>::run(llvm::Module&, llvm::AnalysisManager<llvm::Module>&) (build-all/bin/opt+0x43a39fa)
#19 0x000055ee9ea2d1fb llvm::runPassPipeline(llvm::StringRef, llvm::Module&, llvm::TargetMachine*, llvm::TargetLibraryInfoImpl*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::ToolOutputFile*, llvm::StringRef, llvm::ArrayRef<llvm::PassPlugin>, llvm::ArrayRef<std::function<void (llvm::PassBuilder&)>>, llvm::opt_tool::OutputKind, llvm::opt_tool::VerifierKind, bool, bool, bool, bool, bool, bool, bool) (build-all/bin/opt+0x54cd1fb)
#20 0x000055ee9d6c2bbd optMain (build-all/bin/opt+0x4162bbd)
#21 0x00007f128bf04d85 __libc_start_main (/lib64/libc.so.6+0x3ad85)
#22 0x000055ee9d6bc72e _start (build-all/bin/opt+0x415c72e)
Abort (core dumped)
```
[bbi-98907.ll.gz](https://github.com/user-attachments/files/16927511/bbi-98907.ll.gz)

**Comment by mikaelholmen** [comment]
> Hi, I've no idea what to be expected but I noticed that
> 
> ```
> opt -passes=sandbox-vectorizer bbi-98907.ll -o /dev/null
> ```
> 
> crashes with
> 
> ```
> opt: ../lib/SandboxIR/SandboxIR.cpp:2542: llvm::sandboxir::Value *llvm::sandboxir::Context::registerValue(std::unique_ptr<Value> &&): Assertion `Pair.second && "Already exists!"' failed.
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.

And this crash goes away with d14a600b1eb
```
[SandboxIR] Implement BlockAddress (#107940)
```

**Comment by vporpo** [comment]
> And this crash goes away with https://github.com/llvm/llvm-project/commit/d14a600b1eb650f05fcd56a7b790e30f1f52e751 [SandboxIR] Implement BlockAddress (#107940)

@mikaelholmen Just curious how did you run into that issue?

**Comment by mikaelholmen** [comment]
> > And this crash goes away with [d14a600](https://github.com/llvm/llvm-project/commit/d14a600b1eb650f05fcd56a7b790e30f1f52e751) [SandboxIR] Implement BlockAddress (#107940)
> 
> @mikaelholmen Just curious how did you run into that issue?

We run some tests with randomized opt pass pipelines, and when sandbox-scalarizer was created we threw that pass into the mix as well.

**Comment by vporpo** [comment]
> We run some tests with randomized opt pass pipelines, and when sandbox-vectorizer was created we threw that pass into the mix as well.

Ah makes sense, thanks.

**Comment by gulfemsavrun** [comment]
We started seeing a `CMake` error after this patch.
```
CMake Error: install(EXPORT "LLVMExports" ...) includes target "LLVMVectorize" which requires target "LLVMSandboxIR" that is not in any export set.
```
https://luci-milo.appspot.com/ui/p/fuchsia/builders/prod/llvm-linux-arm64/b8737544888773099041/overview

**Comment by vporpo** [comment]
> We started seeing a CMake error after this patch.

Isn't this a downstream issue?

**Comment by gulfemsavrun** [comment]
> > We started seeing a CMake error after this patch.
> 
> Isn't this a downstream issue?

Ok, you're right, and I'll fix it on our end. Sorry for the false alarm.

**Pull request merged**